src/modm/platform/core/rp/flash.ld.in

```diff
@@ -0,0 +1,90 @@
+%% import "../cortex/linker.macros" as linker with context
+{{ linker.copyright() }}
+
+{{ linker.prefix() }}
+%% set table_heap = []
+%% set table_copy = []
+%% set table_zero = []
+
+%% if vector_table_location == "ram"
+/* Computes stack offset so that vector table is aligned */
+__stack_offset = ALIGN(TOTAL_STACK_SIZE, (1 << LOG2CEIL({{ highest_irq + 16 }})) * 8) - TOTAL_STACK_SIZE;
+%#
+%% endif
+
+SECTIONS
+{
+{{ linker.section_rom_start("FLASH") }}
+
+.boot2 : {
+        __boot2_start__ = .;
+        KEEP (*(.boot2))
+        __boot2_end__ = .;
+    } > FLASH
+ASSERT(__boot2_end__ - __boot2_start__ == 256,
+        "ERROR: Pico second stage bootloader must be 256 bytes in size")
+
+.text : {
+        __vector_table_rom_start = .;
+        __vector_table_ram_load = .;
+        KEEP(*(.vector_rom))
+        __vector_table_rom_end = .;
+
+        KEEP (*(.reset))
+    } > FLASH
+
+{{ linker.section_rom("FLASH") }}
+
+%% macro section_stack_core1(memory, start=None)
+    /* Core1 stack in {{memory}} */
+    .stack (NOLOAD) :
+    {
+    %% if start != None
+        . += {{ start }};
+    %% endif
+        __stack1_start = .;
+
+        . += MAIN_STACK_SIZE;
+        . = ALIGN(8);
+        __core1_stack_top = .;
+
+        . += PROCESS_STACK_SIZE;
+        . = ALIGN(8);
+        __process1_stack_top = .;
+
+        __stack1_end = .;
+    } >{{memory}}
+%% endmacro
+
+{{ linker.section_stack("SCRATCH_Y", "__stack_offset" if vector_table_location == "ram" else None) }}
+{{  section_stack_core1("SCRATCH_X", "__stack1_offset" if vector_table_location == "ram" else None) }}
```

Comment:
Yeah, this is the interesting part, how to assign the memories between the CPUs?

I think it needs to be two lbuild applications with `modm:target=rp2040@0` for CPU0 and `modm:target=rp2040@1` for CPU1 that generate slightly different linkerscripts based on (shared) memory partion settings. Like how much of SRAM0/1/2/3 goes to CPU0 vs CPU1 for example, I guess SRAM4/5 could be used for the main stack and perhaps vector table in ram (4kB may be too small for all use-cases).

For the flash layout: CPU0 has `.boot2` so must go at the bottom, CPU1 then needs to use [the `modm:platform:cortex-m:linkerscript.flash_offset` option](https://modm.io/reference/module/modm-platform-cortex-m/#linkerscriptflash_offset) to move the firmware up above CPU0.
The build system would first build CPU0 to know that offset, then build CPU1 and the combine them into one image?
Seems a bit inelegant and ugly to implement it that way, but we can always fix it later. Would be nice to have some magic that extracts what memories CPU0 uses via pyelftools and then builds the memory map for CPU1 from that, but that's probably not trivial to implement in SCons/Make/CMake.

Reply:
As described at modm-devices, rp2040 dnt need separate compilation for cores, memory is shared with cores. Only separate stack needed.